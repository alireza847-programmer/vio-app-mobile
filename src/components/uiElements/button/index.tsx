import React from 'react';
import Button from './style';
import {VButtonProps} from 'types/components/uiElements/button';
import VText from '../text';
import {theme} from 'themes/emotion';
import VRow from '../row';

const VButton = (props: VButtonProps) => {
  const {
    mode = 'primary',
    subTitle,
    title,
    styled = 'FULL_WIDTH',
    icon,
    rightIcon,
    onPress,
    testID,
    disabled = false,
    width,
  } = props;

  const renderContentColor = () => {
    if (disabled) {
      return theme.colors.textColor.button.disabled;
    }
    if (styled === 'TEXT') {
      return theme.colors.button[mode];
    }
    return theme.colors.textColor.button[mode];
  };

  return (
    <Button
      width={width}
      disabled={disabled}
      testID={testID}
      activeOpacity={0.8}
      onPress={onPress}
      styled={styled}
      mode={mode}>
      {icon && (
        <VRow {...(title && {marginRighRatio: 2})} fullWidth={false}>
          {icon(renderContentColor())}
        </VRow>
      )}
      <VText
        typography="semiBold16"
        color={renderContentColor()}
        marginTopRatio={0}>
        {title}
      </VText>
      {subTitle && (
        <VText
          typography="regular16"
          color={renderContentColor()}
          marginLeftRatio={1.2}
          marginTopRatio={0}>
          {subTitle}
        </VText>
      )}
      {rightIcon && (
        <VRow {...(title && {marginLeftRatio: 2})} fullWidth={false}>
          {rightIcon(renderContentColor())}
        </VRow>
      )}
    </Button>
  );
};
export default VButton;
