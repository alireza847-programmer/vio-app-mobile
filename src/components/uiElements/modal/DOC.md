# VModal Component Documentation

## Overview

The `VModal` component provides a flexible and customizable modal interface for React Native applications. It is built on top of the `react-native-modal` library and includes options for adding headers, close buttons, and custom content.

## Usage

```jsx
import React, {useState} from 'react';
import VModal from './VModal'; // Adjust the import path based on your project structure
import {View, Text} from 'react-native';
import {theme} from 'themes/emotion';

const MyComponent = () => {
  const [isModalVisible, setModalVisible] = useState(false);

  const toggleModal = () => {
    setModalVisible(!isModalVisible);
  };

  return (
    <View>
      <VModal
        isVisible={isModalVisible}
        onClose={toggleModal}
        title="My Modal"
        withClose
        withHeader>
        {/* Your custom content here */}
        <Text>This is the content of the modal.</Text>
      </VModal>

      <Button title="Open Modal" onPress={toggleModal} />
    </View>
  );
};

export default MyComponent;
```

## Props

### VModal Props

- `isVisible` (boolean, required): Controls the visibility of the modal.
- `onClose` (function, required): Function to be called when the modal is closed.
- `title` (string): The title displayed in the modal header.
- `withClose` (boolean): If `true`, a close button is displayed in the header. Default is `false`.
- `withHeader` (boolean): If `true`, a header is displayed with an optional close button and title. Default is `true`.

## Style

The styling for the `VModal` component is defined in the `style` module. It uses Emotion's styling system for enhanced readability and maintainability.

```jsx
import styled, {css} from '@emotion/native';
import VRow from '../row';

export const modalStyle = css({
  justifyContent: 'flex-end',
  alignItems: 'center',
  margin: 0,
  padding: 0,
  marginBottom: 10,
});

export const Wrapper = styled.View(({theme}) => ({
  backgroundColor: theme.colors.backgroundColor,
  justifyContent: 'center',
  alignItems: 'center',
  borderRadius: theme.radius * 2,
  width: '95%',
}));

export const HeaderContainer = styled(VRow)(({theme}) => ({
  borderBottomWidth: 1,
  borderBottomColor: theme.colors.stroke.divider,
}));
```
