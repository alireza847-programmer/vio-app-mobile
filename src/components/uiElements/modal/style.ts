import styled, {css} from '@emotion/native';
import VRow from '../row';

export const modalStyle = css({
  justifyContent: 'flex-end',
  alignItems: 'center',
  margin: 0,
  padding: 0,
  marginBottom: 10,
});

export const Wrapper = styled.View(({theme}) => ({
  backgroundColor: theme.colors.backgroundColor,
  justifyContent: 'center',
  alignItems: 'center',
  borderRadius: theme.radius * 2,
  width: '95%',
}));

export const HeaderContainer = styled(VRow)(({theme}) => ({
  borderBottomWidth: 1,
  borderBottomColor: theme.colors.stroke.divider,
}));
