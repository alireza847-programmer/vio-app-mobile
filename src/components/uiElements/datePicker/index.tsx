import {theme} from '@storybook/react-native';
import React from 'react';
import {View} from 'react-native';
import Calendar from 'react-native-calendar-range-picker';
import {VDatePickerProps} from 'types/components/uiElements/datePicker';

const VDatePicker = (props: VDatePickerProps) => {
  const {...rest} = props;
  return (
    <View style={{width: '100%'}}>
      <Calendar disabledBeforeToday pastYearRange={0} {...rest} />
    </View>
  );
};

export default VDatePicker;
