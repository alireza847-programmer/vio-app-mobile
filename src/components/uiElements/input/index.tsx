import React from 'react';
import Input from './style';
import {VInputProps} from 'types/components/uiElements/input';
import {theme} from 'themes/emotion';

const VInput = (props: VInputProps) => {
  const {...rest} = props;
  return (
    <Input
      placeholderTextColor={theme.colors.textColor.placeHolder}
      {...rest}
    />
  );
};

export default VInput;
