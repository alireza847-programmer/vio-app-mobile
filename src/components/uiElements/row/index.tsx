import styled from '@emotion/native';
import {PropsWithChildren} from 'react';
import {RowProps} from 'types/components/uiElements/row';

const VRow = styled.View<PropsWithChildren<RowProps>>(
  ({
    theme,
    backgroundColor,
    fullWidth = true,
    justifyContent = 'flex-start',
    alignItems = 'center',
    marginTopRatio = 0,
    marginLeftRatio = 0,
    paddingHorizontalRatio = 0,
    paddingVerticalRatio = 0,
    marginRighRatio = 0,
    minWidth,
  }) => ({
    flexDirection: 'row',
    justifyContent,
    alignItems,
    marginTop: theme.verticalRem * marginTopRatio,
    marginLeft: theme.HorizontalRem * marginLeftRatio,
    marginRight: theme.HorizontalRem * marginRighRatio,
    paddingHorizontal: theme.HorizontalRem * paddingHorizontalRatio,
    paddingVertical: theme.verticalRem * paddingVerticalRatio,
    ...(backgroundColor && {backgroundColor}),
    ...(fullWidth && {width: '100%'}),
    ...(minWidth && {minWidth}),
  }),
);

export default VRow;
