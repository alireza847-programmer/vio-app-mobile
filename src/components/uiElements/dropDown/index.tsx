import React, {Fragment, useEffect, useState} from 'react';
import {VDropDownProps} from 'types/components/uiElements/dropDown';
import {theme} from 'themes/emotion';
import {ChevronSvg} from 'assets/svgs';
import {Container} from './style';
import WheelPicker from 'react-native-wheely';
import VModal from '../modal';
import VText from '../text';
import VRow from '../row';
import VButton from '../button';
import {TouchableOpacity} from 'react-native';

const VDropDown = (props: VDropDownProps) => {
  const {data, placeHolder, onChange, defaultValue} = props;
  const [visibleModal, setVisibleModal] = useState(false);
  const [value, setValue] = useState('');
  const [selectedIndex, setSelectedIndex] = useState(0);
  useEffect(() => {
    if (defaultValue) {
      setValue(String(defaultValue));
      setSelectedIndex(data.indexOf(String(defaultValue)));
    } else {
      setValue(String(placeHolder));
    }
  }, [placeHolder, defaultValue]);
  const onConfirm = () => {
    setVisibleModal(false);
    setValue(data[selectedIndex]);
    onChange(data[selectedIndex]);
  };
  const renderModal = () => {
    return (
      <VModal
        withClose
        withHeader
        title="Child 1 age"
        onClose={() => setVisibleModal(false)}
        isVisible={visibleModal}>
        <WheelPicker
          selectedIndex={selectedIndex}
          options={data}
          onChange={index => setSelectedIndex(index)}
          containerStyle={{width: '90%'}}
          opacityFunction={x => 0.3}
          selectedIndicatorStyle={{
            backgroundColor: theme.colors.button.secondary,
            borderTopWidth: 1,
            borderBottomWidth: 1,
            borderColor: theme.colors.button.primary,
          }}
          visibleRest={1}
          itemStyle={{
            backgroundColor: 'transparent',
          }}
          itemTextStyle={{
            ...theme.fonts.semiBold18,
          }}
        />
        <VRow paddingVerticalRatio={4} justifyContent="center">
          <VButton onPress={onConfirm} width={'90%'} title="CONFIRM" />
        </VRow>
      </VModal>
    );
  };
  return (
    <Fragment>
      <TouchableOpacity
        activeOpacity={0.8}
        onPress={() => setVisibleModal(true)}>
        <Container
          alignItems="center"
          justifyContent="space-between"
          fullWidth={false}>
          <VText
            width={theme.HorizontalRem * 10}
            textAlign="left"
            typography="semiBold16">
            {value}
          </VText>
          <VRow fullWidth={false} marginLeftRatio={3}>
            <ChevronSvg fill={theme.colors.button.primary} />
          </VRow>
        </Container>
      </TouchableOpacity>
      {renderModal()}
    </Fragment>
  );
};
export default VDropDown;
