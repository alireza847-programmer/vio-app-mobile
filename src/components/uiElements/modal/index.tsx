import Modal from 'react-native-modal';
import React, {PropsWithChildren} from 'react';
import {VModalProps} from 'types/components/uiElements/modal';
import {HeaderContainer, Wrapper, modalStyle} from './style';
import {theme} from 'themes/emotion';
import {CloseSvg} from 'assets/svgs';
import VText from '../text';
import VButton from '../button';

const VModal = (props: PropsWithChildren<VModalProps>) => {
  const {children, isVisible, onClose, title, withClose, withHeader} = props;
  return (
    <Modal
      animationIn="slideInUp"
      animationOut={'slideOutDown'}
      style={modalStyle}
      testID="modal"
      hasBackdrop={true}
      onBackButtonPress={onClose}
      onBackdropPress={() => onClose()}
      isVisible={isVisible}>
      <Wrapper>
        {withHeader && (
          <HeaderContainer
            justifyContent="space-between"
            paddingHorizontalRatio={4}
            paddingVerticalRatio={6}>
            {withClose && (
              <VButton
                onPress={onClose}
                testID="close-button"
                styled="TEXT"
                icon={fill => <CloseSvg fill={theme.colors.button.primary} />}
              />
            )}
            <VText typography="semiBold16">{title}</VText>
            <CloseSvg fill={'transparent'} />
          </HeaderContainer>
        )}
        {children}
      </Wrapper>
    </Modal>
  );
};
export default VModal;
