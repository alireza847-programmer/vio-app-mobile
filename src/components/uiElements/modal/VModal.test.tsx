import React, {PropsWithChildren} from 'react';
import {render, fireEvent, act} from '@testing-library/react-native';
import VModal from './index';
import {it, describe, expect, jest} from '@jest/globals';
import {ThemeProvider} from '@emotion/react';
import {theme} from 'themes/emotion';
import VText from '../text';
import {VModalProps} from 'types/components/uiElements/modal';

const Component = (props: PropsWithChildren<VModalProps>) => (
  <ThemeProvider theme={theme}>
    <VModal {...props} />
  </ThemeProvider>
);

describe('VModal component', () => {
  it('matches snapshot with default props', () => {
    const {toJSON} = render(<Component isVisible={true} onClose={() => {}} />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('matches snapshot with header and close button', () => {
    const {toJSON} = render(
      <Component
        isVisible={true}
        onClose={() => {}}
        title="My Modal"
        withClose
        withHeader>
        <VText>This is the content of the modal.</VText>
      </Component>,
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('calls onClose when close button is pressed', () => {
    const onCloseMock = jest.fn();
    const {getByTestId} = render(
      <Component isVisible={true} onClose={onCloseMock} withClose withHeader>
        <VText>This is the content of the modal.</VText>
      </Component>,
    );

    const closeButton = getByTestId('close-button');
    act(() => {
      fireEvent.press(closeButton);
    });

    expect(onCloseMock).toHaveBeenCalled();
  });
});
