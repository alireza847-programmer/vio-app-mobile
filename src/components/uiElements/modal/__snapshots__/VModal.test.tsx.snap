// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VModal component matches snapshot with default props 1`] = `
<Modal
  animationType="none"
  deviceHeight={null}
  deviceWidth={null}
  hardwareAccelerated={false}
  hideModalContentWhileAnimating={false}
  onBackdropPress={[Function]}
  onModalHide={[Function]}
  onModalWillHide={[Function]}
  onModalWillShow={[Function]}
  onRequestClose={[Function]}
  panResponderThreshold={4}
  scrollHorizontal={false}
  scrollOffset={0}
  scrollOffsetMax={0}
  scrollTo={null}
  statusBarTranslucent={false}
  supportedOrientations={
    [
      "portrait",
      "landscape",
    ]
  }
  swipeThreshold={100}
  testID="modal"
  transparent={true}
  visible={true}
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "backgroundColor": "black",
        "bottom": 0,
        "height": 1334,
        "left": 0,
        "opacity": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
        "width": 750,
      }
    }
  />
  <View
    collapsable={false}
    deviceHeight={null}
    deviceWidth={null}
    hideModalContentWhileAnimating={false}
    onBackdropPress={[Function]}
    onModalHide={[Function]}
    onModalWillHide={[Function]}
    onModalWillShow={[Function]}
    panResponderThreshold={4}
    pointerEvents="box-none"
    scrollHorizontal={false}
    scrollOffset={0}
    scrollOffsetMax={0}
    scrollTo={null}
    statusBarTranslucent={false}
    style={
      {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "flex-end",
        "margin": 0,
        "marginBottom": 10,
        "padding": 0,
        "transform": [
          {
            "translateY": 1334,
          },
        ],
      }
    }
    supportedOrientations={
      [
        "portrait",
        "landscape",
      ]
    }
    swipeThreshold={100}
  >
    <View
      style={
        [
          {
            "alignItems": "center",
            "backgroundColor": "#fff",
            "borderRadius": 15,
            "justifyContent": "center",
            "width": "95%",
          },
          undefined,
        ]
      }
    />
  </View>
</Modal>
`;

exports[`VModal component matches snapshot with header and close button 1`] = `
<Modal
  animationType="none"
  deviceHeight={null}
  deviceWidth={null}
  hardwareAccelerated={false}
  hideModalContentWhileAnimating={false}
  onBackdropPress={[Function]}
  onModalHide={[Function]}
  onModalWillHide={[Function]}
  onModalWillShow={[Function]}
  onRequestClose={[Function]}
  panResponderThreshold={4}
  scrollHorizontal={false}
  scrollOffset={0}
  scrollOffsetMax={0}
  scrollTo={null}
  statusBarTranslucent={false}
  supportedOrientations={
    [
      "portrait",
      "landscape",
    ]
  }
  swipeThreshold={100}
  testID="modal"
  transparent={true}
  visible={true}
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "backgroundColor": "black",
        "bottom": 0,
        "height": 1334,
        "left": 0,
        "opacity": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
        "width": 750,
      }
    }
  />
  <View
    collapsable={false}
    deviceHeight={null}
    deviceWidth={null}
    hideModalContentWhileAnimating={false}
    onBackdropPress={[Function]}
    onModalHide={[Function]}
    onModalWillHide={[Function]}
    onModalWillShow={[Function]}
    panResponderThreshold={4}
    pointerEvents="box-none"
    scrollHorizontal={false}
    scrollOffset={0}
    scrollOffsetMax={0}
    scrollTo={null}
    statusBarTranslucent={false}
    style={
      {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "flex-end",
        "margin": 0,
        "marginBottom": 10,
        "padding": 0,
        "transform": [
          {
            "translateY": 1334,
          },
        ],
      }
    }
    supportedOrientations={
      [
        "portrait",
        "landscape",
      ]
    }
    swipeThreshold={100}
  >
    <View
      style={
        [
          {
            "alignItems": "center",
            "backgroundColor": "#fff",
            "borderRadius": 15,
            "justifyContent": "center",
            "width": "95%",
          },
          undefined,
        ]
      }
    >
      <View
        justifyContent="space-between"
        paddingHorizontalRatio={4}
        paddingVerticalRatio={6}
        style={
          [
            {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "space-between",
              "marginLeft": 0,
              "marginRight": 0,
              "marginTop": 0,
              "paddingHorizontal": 30,
              "paddingVertical": 40.019999999999996,
              "width": "100%",
            },
            [
              {
                "borderBottomColor": "#EFF2F6",
                "borderBottomWidth": 1,
              },
              undefined,
            ],
          ]
        }
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": false,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "borderColor": "#BFDAF9",
              "borderRadius": 15,
              "flexDirection": "row",
              "justifyContent": "center",
              "opacity": 1,
            }
          }
          testID="close-button"
        >
          <View
            fullWidth={false}
            style={
              [
                {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "justifyContent": "flex-start",
                  "marginLeft": 0,
                  "marginRight": 0,
                  "marginTop": 0,
                  "paddingHorizontal": 0,
                  "paddingVertical": 0,
                },
                undefined,
              ]
            }
          >
            <SvgMock
              fill="#0071F3"
            />
          </View>
          <Text
            color="#0071F3"
            marginTopRatio={0}
            style={
              [
                {
                  "color": "#0071F3",
                  "fontFamily": "Inter-SemiBold",
                  "fontSize": 16,
                  "marginLeft": 0,
                  "marginTop": 0,
                  "textAlign": "center",
                  "textAlignVertical": "center",
                },
                undefined,
              ]
            }
            typography="semiBold16"
          />
        </View>
        <Text
          style={
            [
              {
                "color": "#2A333D",
                "fontFamily": "Inter-SemiBold",
                "fontSize": 16,
                "marginLeft": 0,
                "marginTop": 0,
                "textAlign": "center",
                "textAlignVertical": "center",
              },
              undefined,
            ]
          }
          typography="semiBold16"
        >
          My Modal
        </Text>
        <SvgMock
          fill="transparent"
        />
      </View>
      <Text
        style={
          [
            {
              "color": "#2A333D",
              "fontFamily": "Inter-Medium",
              "fontSize": 15,
              "marginLeft": 0,
              "marginTop": 0,
              "textAlign": "center",
              "textAlignVertical": "center",
            },
            undefined,
          ]
        }
      >
        This is the content of the modal.
      </Text>
    </View>
  </View>
</Modal>
`;
